src/fetch/rewards.py

```diff
@@ -0,0 +1,167 @@
+        .sum()
+    )
+    primary_rewards_df["primary_reward_cow"] = (
+        primary_rewards_df["primary_reward_eth"]
+        .apply(converter.eth_to_token)
+        .astype(object)
+    )
+    return primary_rewards_df
+
+
+def compute_primary_reward_datum(row: Series) -> tuple[str, int]:
+    """Compute the reward of an auction from the winning and references scores"""
+    solver = row["winning_solver"]
+    if row["tx_hash"] is None:
+        reward = -min(row["reference_score"], EPSILON_LOWER)
+    else:
+        reward = min(row["winning_score"] - row["reference_score"], EPSILON_UPPER)
+
+    return solver, reward
+
+
+def compute_secondary_rewards_df(
+    batch_data_df: DataFrame, primary_rewards_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute secondary rewards from batch data and primary rewards"""
+    secondary_allocation = max(
+        min(
+            PERIOD_BUDGET_COW - primary_rewards_df["primary_reward_cow"].sum(),
+            converter.eth_to_token(CONSISTENCY_REWARD_CAP_ETH),
+        ),
+        0,
+    )
+    secondary_rewards_df = (
+        batch_data_df["participating_solvers"]
+        .explode()
+        .value_counts()
+        .reset_index()
+        .rename(columns={"participating_solvers": "solver", "count": "num_auctions"})
+        .astype(object)
+    )
+    participation_total = secondary_rewards_df["num_auctions"].sum()
+    secondary_rewards_df["secondary_reward_cow"] = (
+        secondary_allocation * secondary_rewards_df["num_auctions"]
+    ) // participation_total
+    secondary_rewards_df["secondary_reward_eth"] = (
+        secondary_rewards_df["secondary_reward_cow"]
+        .apply(converter.token_to_eth)
+        .astype(object)
+    )
+    secondary_rewards_df = secondary_rewards_df.drop("num_auctions", axis=1)
+
+    return secondary_rewards_df
+
+
+def compute_quote_rewards(
+    trade_data_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute quote rewards from batch data"""
+    quote_rewards_df = (
+        trade_data_df[trade_data_df.apply(is_market_order, axis=1)]
+        .quote_solver.value_counts()
+        .reset_index()
+        .rename(columns={"quote_solver": "solver", "count": "num_quotes"})
+        .apply(
+            lambda row: (
+                row["solver"],
+                int(
+                    float(
+                        min(
+                            QUOTE_REWARD_COW,
+                            converter.eth_to_token(QUOTE_REWARD_CAP_ETH),
+                        )
+                    )
+                    * row["num_quotes"]
+                ),
+            ),
+            axis=1,
+            result_type="expand",
+        )
+        .rename(columns={0: "solver", 1: "quote_reward_cow"})
+        .astype(object)
+    )
+    return quote_rewards_df
+
+
+def is_market_order(row: Series) -> bool:
+    """Check if an order was in market when created."""
+    if row["quote_solver"] is None:
```

Comment:
Could also rewrite it as 

`if row["quote_solver"] is None or not row["partially_fillable"]:`

and then remove the `not row["partially_fillable"]` condition from the two cases below

Reply:
Good call.

Reply:
Good call.